sql/distsql: TestWindowFunctionsAgainstProcessor failed

sql/distsql.TestWindowFunctionsAgainstProcessor [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3249385&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3249385&tab=artifacts#/) on master @ [31af9e32a55a166166e9ba9c5327b7cd847ae236](https://github.com/cockroachdb/cockroach/commits/31af9e32a55a166166e9ba9c5327b7cd847ae236):


```
=== RUN   TestWindowFunctionsAgainstProcessor
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWindowFunctionsAgainstProcessor917519800
    test_log_scope.go:74: use -show-logs to present logs inline
window function: MAX
partitionCols: [0]
ordering: 1 DESC, 
argIdxs: [1]
frame mode: RANGE
start bound: {OFFSET_PRECEDING 0 [4 64 8 0 0 0 0 0 0] {DESCENDING_KEY float4}}
end bound: {UNBOUNDED_FOLLOWING 0 [4 64 34 0 0 0 0 0 0] {DESCENDING_KEY float4}}
frame exclusion: NO_EXCLUSION
seed = 7475079935636293304
CREATE TABLE t(a bigint, b real);
INSERT INTO t VALUES
(3443419988752884857, 0.11126037687063217),
(NULL, -0.6579824686050415),
(-7885120243327823855, 0.48770201206207275),
(2749514120986619359, 0.9637739062309265),
(2503731026000987132, +Inf),
(NULL, NULL),
(NULL, -0.5484631061553955),
(-1536058271231931927, -0.44659504294395447),
(3104309310026389454, NULL),
(8857656624477695434, -2.340207815170288);
    columnar_operators_test.go:1198: different results on row 9;
        expected:
        [NULL NULL -0.6579824686050415]
        got:
        [NULL NULL NULL]
    panic.go:632: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestWindowFunctionsAgainstProcessor917519800
--- FAIL: TestWindowFunctionsAgainstProcessor (1.87s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestWindowFunctionsAgainstProcessor PKG=./pkg/sql/distsql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-queries yuzefovich
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWindowFunctionsAgainstProcessor.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/window
+++ b/pkg/sql/logictest/testdata/logic_test/window
@@ -3961,3 +3961,22 @@ SELECT first_value(x) OVER (ORDER BY x RANGE BETWEEN CURRENT ROW AND '0 YEAR'::I
 
 statement ok
 RESET vectorize;
+
+# Regression test for incorrect results for min and max when the window frame
+# shrinks.
+statement ok
+DROP TABLE IF EXISTS t;
+CREATE TABLE t (a INT);
+INSERT INTO t VALUES (1), (-1), (NULL);
+SET vectorize=off;
+
+query III rowsort
+SELECT a, max(a) OVER w, min(a) OVER w FROM t
+WINDOW w AS (ORDER BY a DESC RANGE BETWEEN 10 PRECEDING AND UNBOUNDED FOLLOWING);
+----
+1     1     -1
+-1    1     -1
+NULL  NULL  NULL
+
+statement ok
+RESET vectorize;

--- a/pkg/sql/sem/builtins/window_frame_builtins.go
+++ b/pkg/sql/sem/builtins/window_frame_builtins.go
@@ -69,10 +69,7 @@ func (sw *slidingWindow) add(iv *indexedValue) {
 // indices smaller than given 'idx'. This operation corresponds to shifting the
 // start of the frame up to 'idx'.
 func (sw *slidingWindow) removeAllBefore(idx int) {
-	for i := 0; i < sw.values.Len() && i < idx; i++ {
-		if sw.values.Get(i).(*indexedValue).idx >= idx {
-			break
-		}
+	for sw.values.Len() > 0 && sw.values.Get(0).(*indexedValue).idx < idx {
 		sw.values.RemoveFirst()
 	}
 }

